src/System/extract_state.rs

```diff
@@ -73,36 +73,41 @@ fn build_location_tuple(
     machine: &SystemRecipe,
     system: &TransitionSystemPtr,
 ) -> Result<LocationTuple, String> {
-    let location_id = get_location_id(&mut locations.iter(), machine);
+    let location_id = get_location_id(&mut locations.iter(), machine)?;
     let locations_system = system.get_all_locations();
-    let locationtuple = locations_system.iter().find(|loc| loc.id == location_id);
-
-    if locationtuple.is_none() {
-        return Err(format!(
-            "The location {} is not found in the system",
-            location_id
-        ));
+    if let Some(locationtuple) = locations_system.iter().find(|loc| loc.id == location_id) {
+        Ok(locationtuple.clone())
+    } else {
+        Ok(LocationTuple::create_partial_location(location_id))
     }
-
-    Ok(locationtuple.unwrap().clone())
 }
 
-fn get_location_id(locations: &mut Iter<&str>, machine: &SystemRecipe) -> LocationID {
+fn get_location_id(
+    locations: &mut Iter<&str>,
+    machine: &SystemRecipe,
+) -> Result<LocationID, String> {
     match machine {
-        SystemRecipe::Composition(left, right) => LocationID::Composition(
-            Box::new(get_location_id(locations, left)),
-            Box::new(get_location_id(locations, right)),
-        ),
-        SystemRecipe::Conjunction(left, right) => LocationID::Conjunction(
-            Box::new(get_location_id(locations, left)),
-            Box::new(get_location_id(locations, right)),
-        ),
-        SystemRecipe::Quotient(left, right, _clock_index) => LocationID::Quotient(
-            Box::new(get_location_id(locations, left)),
-            Box::new(get_location_id(locations, right)),
-        ),
-        SystemRecipe::Component(_comp) => {
-            LocationID::Simple(locations.next().unwrap().trim().to_string())
-        }
+        SystemRecipe::Composition(left, right) => Ok(LocationID::Composition(
+            Box::new(get_location_id(locations, left)?),
+            Box::new(get_location_id(locations, right)?),
+        )),
+        SystemRecipe::Conjunction(left, right) => Ok(LocationID::Conjunction(
+            Box::new(get_location_id(locations, left)?),
+            Box::new(get_location_id(locations, right)?),
+        )),
+        SystemRecipe::Quotient(left, right, _clock_index) => Ok(LocationID::Quotient(
+            Box::new(get_location_id(locations, left)?),
+            Box::new(get_location_id(locations, right)?),
+        )),
+        SystemRecipe::Component(comp) => match locations.next().unwrap().trim() {
+            "_" => Ok(LocationID::AnyLocation()),
+            str => {
+                if comp.get_locations().iter().any(|loc| loc.get_id() == str) {
```

Comment:
this is suddenly much more expensive, is this strictly necessary?
Instead of O(n) it is now O(N * M)
Where N is the amount of simples and M is the amount of locations in the machine

Reply:
Nah, fixed